src/dataset.cpp

```diff
@@ -574,8 +575,183 @@ Dataset concatenate(const Dataset &d1, const Dataset &d2, const Dim dim) {
+    MDZipView<const Coord::Position> specPos(d);
+    std::transform(specPos.begin(), specPos.end(),
+                   scale.get<Data::Value>().begin(), [&](const auto &item) {
+                     const auto &pos = item.template get<Coord::Position>();
+                     const double l2 = (samplePos - pos).norm();
+                     return l2 * l2;
+                   });
+  } else if (d.contains(Coord::Ef{})) {
+    // Indirect-inelastic.
+
+    tofShift.setDimensions(dims);
+    // Ef can be different for every spectrum so we access it also via a view.
+    MDZipView<const Coord::Position, const Coord::Ef> geometry(d);
+    std::transform(geometry.begin(), geometry.end(),
+                   tofShift.get<Data::Value>().begin(), [&](const auto &item) {
+                     const auto &pos = item.template get<Coord::Position>();
+                     const auto &Ef = item.template get<Coord::Ef>();
+                     const double l2 = (samplePos - pos).norm();
+                     return l2 * l2 / sqrt(Ef);
+                   });
+
+    scale.setDimensions({});
+    scale.get<Data::Value>()[0] = l1 * l1;
+  } else {
+    throw std::runtime_error("Dataset contains neither Coord::Ei nor "
+                             "Coord::Ef, this does not look like "
+                             "inelastic-scattering data.");
+  }
+
+  // 2. Transform variables
+  Dataset converted;
+  for (const auto &var : d) {
+    auto varDims = var.dimensions();
+    if (varDims.contains(Dim::Tof))
+      varDims.relabel(varDims.index(Dim::Tof), Dim::DeltaE);
+    if (var.tag() == Coord::Tof{}) {
+      auto dims = scale.dimensions();
+      for (const Dim dim : varDims.labels())
+        if (!dims.contains(dim))
+          dims.addInner(dim, varDims[dim]);
+      Variable E(Coord::DeltaE{}, dims, dims.volume(), 1.0);
+      E *= var.reshape(varDims);
+      E -= tofShift;
+      E *= E;
+      E = 1.0 / std::move(E);
+      E *= scale;
+      if (d.contains(Coord::Ei{})) {
+        converted.insert(-(std::move(E) - d(Coord::Ei{})));
+      } else {
+        converted.insert(std::move(E) - d(Coord::Ef{}));
+      }
+    } else if (var.tag() == Data::Events{}) {
+      throw std::runtime_error(
+          "TODO Converting units of event data not implemented yet.");
+    } else {
+      converted.insert(var.reshape(varDims));
+    }
+  }
+
+  return converted;
+}
+} // namespace tof
+} // namespace neutron
+
 Dataset convert(const Dataset &d, const Dim from, const Dim to) {
-  static_cast<void>(to);
+  if ((from == Dim::Tof) && (to == Dim::Energy))
+    return neutron::tof::tofToEnergy(d);
+  if ((from == Dim::Tof) && (to == Dim::DeltaE))
+    return neutron::tof::tofToDeltaE(d);
+  throw std::runtime_error(
+      "Conversion between requested dimensions not implemented yet.");
   // How to convert? There are several cases:
   // 1. Tof conversion as Mantid's ConvertUnits.
   // 2. Axis conversion as Mantid's ConvertSpectrumAxis.
```

Comment:
I'm not sure we need to make a distinction. The whole vertical axis thing seems tied up with the Workspace2D structure. Isn't this now just a case of what conversions are defined possible. i.e `spectrum_position` -> `two_theta`, `tof->wavelength`

Reply:
Exactly, that was also my idea. But for the spectrum axis there is a slightly complication: It essentially starts off as a discrete, non-continuous axis, but can be converted into one.

It is also not so clear of for `ConvertSpectrumAxis` we actually *want* to convert the dimension: In the `Dataset` way of operation, all data would then not have `Dim::Spectrum` but `Dim::TwoTheta` (or similar). It thus cannot be indexed by spectrum number any more, etc., ... The alternative would be to simply add a new coordinate `Coord::TwoTheta` but with `Dim::Spectrum` (*not* `Dim::TwoTheta`).

Reply:
On the other hand, maybe we should have something like `Dim::Position` as the common "primary" dimension, instead of `Dim::Spectrum`? The latter is after all just the dimension we use as a matter of habit, not because it is the natural representation of data? We could of course still have `Coord::SpectrumNumber`, but it would have `Dim::Position`.

Would this description work also well for other data where spectra are not linked to positions? For example, after `SofQw` we would have `Dim::Q` (instead of `Dim::Spectrum`). The disadvantage I can see is that data now has a different dimension label depending on what it represents. But maybe that is actually an advantage? After all I am proposing to do the same for `Dim::Tof` which would be something else after unit conversion.

Reply:
I agree that a `Dim::Position` makes more sense to me than `Dim::Spectrum` which is confusing to anyone new to Mantid. It also makes it more general, if `Dataset` is ever to be applied to something other than neutrons or Mantid. But then you also have `Coord::Position`, so this might make things more confusing...

Reply:
`Dim::Position` and `Coord::Position` are linked.

Regarding confusing people new to Mantid: I do not know if `Dim::Position` is less confusing, you can ask "position of what"?